File: trunk/compact-net/CompactNetMercFramework/Merc.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Windows.Forms;
using System.Reflection;

namespace it.mintlab.mobilenet.mercframework
{
    public class Merc : IMerc
    {
        protected IMDispatcher dispatcher;
        private Dictionary<string, MethodInfo> stateBindingMap; // stateName <-> MethodInfo
        private string state = StateBinding.DEFAULT_STATE;
        private bool onStateChanging = false;

        /// <summary>
        /// State Property for subclasses
        /// </summary>
        /// <returns></returns>
        protected string State
        {
            get { return state.Substring(1); }
            set
            {
                value = StateBinding.STATE_PREFIX + value;
                if (!state.Equals(value))
                {
                    if (onStateChanging)
                    {
                        dispatcher.logError("Change to " + value + " is not allowed during change from " + state, null);
                        return;
                    }
                    onStateChanging = true;
                    if (stateBindingMap.ContainsKey(value))
                    {
                        // State method invocation
                        stateBindingMap[value].Invoke(this, null);
                    }
                    dispatcher.logInfo("State change: " + state + " -> " + value);
                    state = value;
                    onStateChanging = false;
                }
                else
                    dispatcher.logInfo("Same State " + state + " change attempted");
            }
        }

        /// <summary>
        /// Dispatcher Setter
        /// </summary>
        /// <param id="dispatcher">Merc dispatcher</param>
        public void setDispatcher(IMDispatcher dispatcher)
        {
            this.dispatcher = dispatcher;
        }

        /// <summary>
        /// Get the state of Merc
        /// </summary>
        /// <returns></returns>
        public string getState() { return state; }

        /// <summary>
        /// Set Binding beetween state and method to call for state changing
        /// </summary>
        public void loadStateBinding()
        {
            stateBindingMap = new Dictionary<string, MethodInfo>();
            MethodInfo[] methods = this.GetType().GetMethods();
            foreach (MethodInfo mi in methods)
            {
                Attribute[] attrs = Attribute.GetCustomAttributes(mi, typeof(StateBinding), true);
                foreach (Attribute a in attrs)
                {
                    string state = StateBinding.STATE_PREFIX + ((StateBinding)a).state;
                    if (!stateBindingMap.ContainsKey(state))
                    {
                        stateBindingMap.Add(state, mi);
                    }
                    else
                    {
                        dispatcher.logError("Method already present for state " + state, null);
                    }
                }
            }
        }

    }
}

File: java/JavaMercFramework/src/main/it/mintlab/desktopjava/mercframework/Merc.java
package it.mintlab.desktopjava.mercframework;

import java.lang.annotation.Annotation;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;

	public class Merc implements IMerc
	{
		protected IMDispatcher dispatcher;
		private HashMap<String, Method> stateBindingMap; // stateName <-> MethodInfo
		private String state = StateBinding.DEFAULT_STATE;
		private boolean onStateChanging = false;

		/** 
		* State property for subclasses
		*/
		public final String getState()
		{
			//return state.substring(1);
			return state;
		}
		
		protected final void setState(String value)
		{
			if(onStateChanging)
			{
				dispatcher.logError("Change to "+value+" is not allowed during change from "+state);
				return;
			}
			onStateChanging = true;
			//value = StateBinding.STATE_PREFIX + value;
			if (state.equals(value)) return;
			if (stateBindingMap.containsKey(value))
			{
				// State method invocation
				try {
					stateBindingMap.get(value).invoke(this, null);
				} catch (Exception e) {
					dispatcher.logError(value + " State Method Invocation Exception");
				}
			}
			dispatcher.logInfo("State change: " + state + " -> " + value);
			state = value;
			onStateChanging = false;
		}

		/* 
		 Get the state of Merc
		 
		 @return 
		
		public final String getRealState()
		{
			return state;
		}*/

		/** 
		 Dispatcher Setter
		 
		 <param id="dispatcher">Merc dispatcher
		*/
		public final void setDispatcher(IMDispatcher dispatcher)
		{
			this.dispatcher = dispatcher;
		}

		/** 
		* Set Binding beetween state and method to call for state changing
		*/
		public final void loadStateBinding()
		{
			stateBindingMap = new HashMap<String, Method>();
			Method[] methods = this.getClass().getDeclaredMethods();
			for(Method m : methods)
			{
				Annotation[] annotations = m.getAnnotations();
				for(Annotation a : annotations)
				{
					if(a.annotationType().equals(StateBinding.class))
					{	
						StateBinding sb = (StateBinding)a;
						//String state = StateBinding.STATE_PREFIX + sb.state();
						String state = sb.state();
						if (!stateBindingMap.containsKey(state))
						{
							stateBindingMap.put(state, m);
						}
						else
						{
							dispatcher.logError("Method already present for state " + state);
						}
					}
					
				}
			}
		}
	}